lib/Backends/OpenCL/OpenCLDeviceManager.cpp

```diff
@@ -181,6 +181,52 @@ Error OpenCLDeviceManager::parseConfig() {
   return Error::success();
 }
 
+Error OpenCLDeviceManager::findBestDevice(cl_platform_id platformId,
+                                          int deviceId) {
+  cl_uint num{0};
+  cl_int err = clGetDeviceIDs(platformId, CL_DEVICE_TYPE_ALL, 0, nullptr, &num);
+  if (err != CL_SUCCESS) {
+    RETURN_ERR("clGetDeviceIDs Failed");
+  }
+  if ((deviceId > 0 && num < deviceId) || num == 0) {
+    RETURN_ERR("Should have at least one GPU/CPU/FPGA for running OpenCL");
+  }
+
+  std::vector<cl_device_id> devices(num);
+  err = clGetDeviceIDs(platformId, CL_DEVICE_TYPE_ALL, num, devices.data(),
+                       nullptr);
+  if (err != CL_SUCCESS) {
+    RETURN_ERR("clGetDeviceIDs Failed");
+  }
+
+  if (deviceId >= 0) {
+    deviceId_ = devices[deviceId];
+  } else {
+    cl_device_id chosen = devices[0];
```

Comment:
Could you add some comments about the logic in the else-branch? You query a more detailed info about each device and then you play with the chosen_type. It would be nice to have some comments about this.